tomviz/ExternalPythonExecutor.cxx

```diff
@@ -0,0 +1,166 @@
+/* This source file is part of the Tomviz project, https://tomviz.org/.
+   It is released under the 3-Clause BSD License, see "LICENSE". */
+
+#include "ExternalPythonExecutor.h"
+#include "DataSource.h"
+#include "EmdFormat.h"
+#include "Operator.h"
+#include "OperatorPython.h"
+#include "Pipeline.h"
+#include "PipelineWorker.h"
+
+namespace tomviz {
+
+ExternalPythonExecutor::ExternalPythonExecutor(Pipeline* pipeline)
+  : ExternalPipelineExecutor(pipeline)
+{
+}
+
+ExternalPythonExecutor::~ExternalPythonExecutor() = default;
+
+Pipeline::Future* ExternalPythonExecutor::execute(vtkDataObject* data,
+                                                  QList<Operator*> operators,
+                                                  int start, int end)
+{
+
+  auto future = ExternalPipelineExecutor::execute(data, operators, start, end);
+
+  // We are now ready to run the pipeline
+  QStringList args = executorArgs(start);
+
+  PipelineSettings settings;
+  auto pythonExecutable = settings.externalPythonExecutablePath();
+
+  // Find the tomviz-pipeline executable
+  auto pythonExecutableFile = QFileInfo(pythonExecutable);
+
+  if (!pythonExecutableFile.exists()) {
+    displayError("External Python Error",
+                 QString("The external python executable doesn't exist: %1\n")
+                   .arg(pythonExecutable));
+
+    return Pipeline::emptyFuture();
+  }
+
+  auto baseDir = pythonExecutableFile.dir();
+  auto tomvizPipelineExecutable =
+    QFileInfo(baseDir.filePath("tomviz-pipeline"));
+  if (!tomvizPipelineExecutable.exists()) {
+    displayError(
+      "External Python Error",
+      QString("Unable to find tomviz-pipeline executable, please ensure "
+              "tomviz package has been installed in python environment."));
+
+    return Pipeline::emptyFuture();
+  }
+
+  m_process.reset(new QProcess(this));
+
+  connect(m_process.data(), &QProcess::errorOccurred, this,
+          &ExternalPythonExecutor::error);
+  connect(
+    m_process.data(),
+    QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
+    [this](int exitCode, QProcess::ExitStatus exitStatus) {
+
+      if (exitStatus == QProcess::CrashExit) {
+        displayError("External Python Error",
+                     QString("The external python process crash: %1\n\n "
+                             "stderr:\n%2 \n\n stdout:\n%3 \n")
+                       .arg(commandLine(this->m_process.data()))
+                       .arg(QString(this->m_process->readAllStandardError()))
+                       .arg(QString(this->m_process->readAllStandardOutput())));
```

Comment:
Adding support for reading all standard error and output on successful runs would be very useful too - sometimes the algorithm fails even though the process succeeds and print statements can help debug.

Reply:
Done